pkg/proxy/backend/error.go

```diff
@@ -0,0 +1,75 @@
+// Copyright 2023 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package backend
+
+import (
+	"github.com/pingcap/TiProxy/lib/util/errors"
+	pnet "github.com/pingcap/TiProxy/pkg/proxy/net"
+	"github.com/pingcap/tidb/parser/mysql"
+	"go.uber.org/zap"
+)
+
+const (
+	connectErrMsg    = "No available TiDB instances, please check TiDB cluster"
+	handshakeErrMsg  = "TiProxy fails to connect to TiDB, please check network"
+	capabilityErrMsg = "Verify TiDB capability failed, please upgrade TiDB"
+)
+
+// UserError is returned to the client.
+// err is used to log and userMsg is used to report to the user.
+type UserError struct {
+	err     error
+	userMsg string
+}
+
+func WrapUserError(err error, userMsg string) *UserError {
```

Comment:
I think this is duplicated with `errors.Wrapf(err, "Ggg")`, which is basically the samething. And you can get the msg by `errors.Unwrap(wrapedErr).Error()`.

We just need to define `ErrUserError` to reuse `lib/util/errors`.

Reply:
Do you mean:
To wrap: err = errors.Wrapf(errors.Wrapf(ErrUserError, err), userMsg)
To get userMsg: errors.Unwrap(err).Error()
To get logMsg: err.Error() or errors.Unwrap(errors.Unwrap(err)).Error()

Reply:
I mean:

To wrap: `err = errors.Wrapf(ErrUserError, userMsg)`
To get userMsg: `errors.Unwrap(user_err).Error()`

If it is user error, then it is also safe to log. If it is not, the internal error should be logged by  either gateway or tiproxy. I mean `logMsg` and `userMsg` are the samething for internal errors.

Reply:
The internal error msg like dial timeout, and EOF, are lost?

Reply:
For example, `no instance` and `dial timeout` are both replaced by `ErrUserError`, then when I log the msg, I don't know the original error.
A walkaround is to log the internal errors whenever they are generated, but this brings more code.